File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/transactions/TransactionStateTracker.kt
package com.gemwallet.android.data.repositories.transactions

import android.util.Log
import com.gemwallet.android.cases.transactions.CreateTransaction
import com.gemwallet.android.ext.runCatchingCancellable
import com.gemwallet.android.ext.toIdentifier
import com.gemwallet.android.serializer.decodeJson
import com.gemwallet.android.serializer.toJson
import com.wallet.core.primitives.Asset
import com.wallet.core.primitives.AssetId
import com.wallet.core.primitives.Transaction
import com.wallet.core.primitives.Wallet
import com.wallet.core.primitives.WalletId
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import uniffi.gemstone.GemTransactionStateService

private const val TAG = "TransactionStateTracker"

class TransactionStateTracker(
    private val stateService: GemTransactionStateService,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) : CreateTransaction {

    fun start() {
        trackPendingTransactions()
    }

    fun stop() {
        stateService.stopTracking()
    }

    override fun trackPendingTransactions() {
        scope.launch {
            runCatchingCancellable { stateService.trackPending() }
                .onFailure { Log.e(TAG, "pending transactions tracking failed", it) }
        }
    }

    override suspend fun trackTransactions(walletId: WalletId, transactions: List<Transaction>) {
        track(walletId, transactions)
    }

    override suspend fun createNotificationTransaction(wallet: Wallet, assetId: AssetId, transaction: Transaction): Asset? {
        val asset = stateService.addNotificationTransaction(wallet.toJson(), assetId.toIdentifier(), transaction.toJson())
            ?.decodeJson<Asset>() ?: return null
        track(wallet.id, listOf(transaction))
        return asset
    }

    private fun track(walletId: WalletId, transactions: List<Transaction>) {
        scope.launch {
            runCatchingCancellable { stateService.track(walletId.id, transactions.map { it.toJson() }) }
                .onFailure { Log.e(TAG, "tracking failed for ${transactions.map { it.id.hash }}", it) }
        }
    }
}

File: ios/Packages/GemstoneServices/Sources/TransactionState/TransactionStateTracker.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemTransactionStateServiceProtocol
import GemstonePrimitives
import Primitives

public struct TransactionStateTracker: Sendable {
    private let service: any GemTransactionStateServiceProtocol

    public init(service: any GemTransactionStateServiceProtocol) {
        self.service = service
    }

    public func trackPending() {
        Task {
            do {
                try await service.trackPending()
            } catch {
                debugLog("transaction state: pending tracking failed: \(error)")
            }
        }
    }

    public func track(wallet: Wallet, transactions: [Transaction]) {
        Task {
            do {
                try await service.track(walletId: wallet.id.id, transactions: transactions.map { try $0.json() })
            } catch {
                debugLog("transaction state: tracking failed: \(error)")
            }
        }
    }

    public func addNotificationTransaction(wallet: Wallet, assetId: AssetId, transaction: Transaction) async throws -> Asset? {
        guard let asset = try await service.addNotificationTransaction(
            wallet: wallet.json(),
            assetId: assetId.identifier,
            transaction: transaction.json(),
        ) else {
            return .none
        }
        track(wallet: wallet, transactions: [transaction])
        return try Asset(asset)
    }
}
